[zh] Sync Protect Your Mission-Critical Pods From Eviction With PriorityClass blog into Chinese (#42544)
kubernetes/website: content/zh-cn/blog/_posts/2023-01-12-protect-mission-critical-pods-priorityclass/index.md

---
layout: blog
title: "使用 PriorityClass 确保你的关键任务 Pod 免遭驱逐"
date: 2023-01-12
slug: protect-mission-critical-pods-priorityclass
description: "Pod 优先级和抢占有助于通过决定调度和驱逐的顺序来确保关键任务 Pod 在资源紧缩的情况下正常运行。"
---

<!--
layout: blog
title: "Protect Your Mission-Critical Pods From Eviction With PriorityClass"
date: 2023-01-12
slug: protect-mission-critical-pods-priorityclass
description: "Pod priority and preemption help to make sure that mission-critical pods are up in the event of a resource crunch by deciding order of scheduling and eviction."
-->

**作者**：Sunny Bhambhani (InfraCloud Technologies)

<!--
**Author:** Sunny Bhambhani (InfraCloud Technologies)
-->

> **译者**：Wilson Wu (DaoCloud)

Reviewer: AIUI, we usually don't credit localization teams in the page. We do appreciate the help though.

Author: OK, I saw most of other translation in the blogs have it, if need, I can remove it.

The control plane consists of multiple components, out of which the scheduler (usually the built-in [kube-scheduler](/docs/concepts/scheduling-eviction/kube-scheduler/)) is one of the components which is responsible for assigning a node to a pod.
-->
控制平面由多个组件组成，其中调度程序（通常是内置的 [kube-scheduler](/zh-cn/docs/concepts/scheduling-eviction/kube-scheduler/)）是一个负责为 Pod 分配节点的组件。

<!--
Whenever a pod is created, it enters a "pending" state, after which the scheduler determines which node is best suited for the placement of the new pod.
-->
当 Pod 被创建时，它就会进入“Pending”状态，之后调度程序会确定哪个节点最适合放置这个新 Pod。

<!--
In the background, the scheduler runs as an infinite loop looking for pods without a `nodeName` set that are [ready for scheduling](/docs/concepts/scheduling-eviction/pod-scheduling-readiness/). For each Pod that needs scheduling, the scheduler tries to decide which node should run that Pod.
-->
在后台，调度程序以无限循环的方式运行，并寻找没有设置 `nodeName` 且[准备好进行调度](/zh-cn/docs/concepts/scheduling-eviction/pod-scheduling-readiness/)的 Pod。
对于每个需要调度的 Pod，调度程序会尝试决定哪个节点应该运行该 Pod。

<!--
If the scheduler cannot find any node, the pod remains in the pending state, which is not ideal.
-->
如果调度程序找不到任何节点，Pod 就会保持在这个不理想的挂起状态下。

> <!--
> {{< note >}}
> To name a few, `nodeSelector`, `taints and tolerations`, `nodeAffinity`, the rank of nodes based on available resources (for example, CPU and memory), and several other criteria are used to determine the pod's placement.
> {{< /note >}}
> -->
> {{< note >}}
> 举几个例子，`nodeSelector`、`taints and tolerations`、`nodeAffinity`、
> 基于可用资源（例如 CPU 和内存）的节点排名以及其他几个用于确定 Pod 位置的标准。
> {{< /note >}}

Reviewer 2: ```suggestion
{{< note >}}
<!--
To name a few, `nodeSelector`, `taints and tolerations`, `nodeAffinity`, the rank of nodes based on available resources (for example, CPU and memory), and several other criteria are used to determine the pod's placement.
-->
举几个例子，可以用 `nodeSelector`、`taints and tolerations`、`nodeAffinity`、
基于可用资源（例如 CPU 和内存）排列节点以及其他若干判别标准来确定将 Pod 放到哪个节点。
{{< /note >}}
```

Author: Fixed in the latest update.

-->
此外，当健康集群遇到节点故障时，通常情况下，较低优先级的 Pod 会被抢占，以便在可用节点上为较高优先级的 Pod 创建空间。
即使集群可以自动创建新节点，也会发生这种情况，因为 Pod 创建通常比创建新节点快得多。

<!--
### PriorityClass requirements
-->
### PriorityClass 的前提条件 {#priorityclass-requirements}

<!--
Before you set up PriorityClasses, there are a few things to consider.
-->
在设置 PriorityClass 之前，需要考虑一些事项。

<!--
1. Decide which PriorityClasses are needed. For instance, based on environment, type of pods, type of applications, etc.
2. The default PriorityClass resource for your cluster. The pods without a `priorityClassName` will be treated as priority 0.
3. Use a consistent naming convention for all PriorityClasses.
4. Make sure that the pods for your workloads are running with the right PriorityClass.
-->
1. 决定哪些 PriorityClass 是需要的。例如，基于环境、Pod 类型、应用类型等。
2. 集群中默认的 PriorityClass 资源。当 Pod 没有设置 `priorityClassName` 时，优先级将被视为 0。
3. 对所有 PriorityClass 使用一致的命名约定。
4. 确保工作负载的 Pod 正在使用正确的 PriorityClass。

<!--
## PriorityClass hands-on example
-->
## PriorityClass 的动手示例 {#priorityclass-hands-on-example}

<!--
Let’s say there are 3 application pods: one for prod, one for preprod, and one for development. Below are three sample YAML manifest files for each of those.
-->

> 假设有 3 个应用 Pod：一个用于生产（prod），一个用于预生产（prepord），一个用于开发（development）。下面是这三个示例的 YAML 清单文件。

Reviewer 2: ```suggestion
假设有 3 个应用 Pod：一个用于生产（prod），一个用于预生产（prepord），一个用于开发（development）。
下面是这三个示例的 YAML 清单文件。
```

Author: Fixed in the latest update.

```yaml
---

> # development

Reviewer: ```suggestion
# 开发环境 (dev)
```
?

Author: Translated in the latest update.

apiVersion: v1
kind: Pod
metadata:
  name: dev-nginx
  labels:
    env: dev
spec:
  containers:
  - name: dev-nginx
    image: nginx
    resources:
      requests:
        memory: "256Mi"
        cpu: "0.2"
      limits:
        memory: ".5Gi"
        cpu: "0.5"
```

```yaml
---

> # preproduction

Reviewer: ```suggestion
# 预生产环境 (preprod)
```

Author: Translated in the latest update.

apiVersion: v1
kind: Pod
metadata:
  name: preprod-nginx
  labels:
    env: preprod
spec:
  containers:
  - name: preprod-nginx
    image: nginx
    resources:
      requests:
        memory: "1.5Gi"
        cpu: "1.5"
      limits:
        memory: "2Gi"
        cpu: "2"
```

```yaml
---

> # production

Reviewer: ```suggestion
# 生产环境 (prod)
```

Author: Translated in the latest update.

apiVersion: v1
kind: Pod
metadata:
  name: prod-nginx
  labels:
    env: prod
spec:
  containers:
  - name: prod-nginx
    image: nginx
    resources:
      requests:
        memory: "2Gi"
        cpu: "2"
      limits:
        memory: "2Gi"
        cpu: "2"
```

<!--
You can create these pods with the `kubectl create -f <FILE.yaml>` command, and then check their status using the `kubectl get pods` command. You can see if they are up and look ready to serve traffic:
-->
你可以使用 `kubectl create -f <FILE.yaml>` 命令创建这些 Pod，然后使用 `kubectl get pods` 命令检查它们的状态。
你可以查看它们是否已启动并准备好提供流量：

```console
$ kubectl get pods --show-labels
NAME            READY   STATUS    RESTARTS   AGE   LABELS
dev-nginx       1/1     Running   0          55s   env=dev
preprod-nginx   1/1     Running   0          55s   env=preprod
prod-nginx      0/1     Pending   0          55s   env=prod
```

<!--
Bad news. The pod for the Production environment is still Pending and isn't serving any traffic.
-->

> 坏消息是生产环境的 Pod 仍处于 Pending 状态，并且不能提供任何流量。

Author: Fixed in the latest update.

<!--
Let's see why this is happening:
-->
让我们看看为什么会发生这种情况：

```console
$ kubectl get events
...
...
5s          Warning   FailedScheduling   pod/prod-nginx      0/2 nodes are available: 1 Insufficient cpu, 2 Insufficient memory.
```

<!--
In this example, there is only one worker node, and that node has a resource crunch.
-->
在此示例中，只有一个工作节点，并且该节点存在资源紧缩。

<!--
Now, let's look at how PriorityClass can help in this situation since prod should be given higher priority than the other environments.
-->
现在，让我们看看在这种情况下 PriorityClass 如何提供帮助，因为生产环境应该比其他环境具有更高的优先级。

<!--
## PriorityClass API
-->
## PriorityClass 的 API {#priorityclass-api}

<!--
Before creating PriorityClasses based on these requirements, let's see what a basic manifest for a PriorityClass looks like and outline some prerequisites:
-->
在根据这些需求创建 PriorityClass 之前，让我们看看 PriorityClass 的基本清单是什么样的，并提出一些先决条件：

```yaml
apiVersion: scheduling.k8s.io/v1
kind: PriorityClass
metadata:
  name: PRIORITYCLASS_NAME
value: 0 # any integer value between -1000000000 to 1000000000 
description: >-
  (Optional) description goes here!

> globalDefault: false # or true. Only one PriorityClass can be the global default.

Reviewer: Consider localizing the comments and the sample description.

Author: Translated in the latest update.

  - `globalDefault`: When true, this PriorityClass is used for pods where a `priorityClassName` is not specified. Only one PriorityClass with `globalDefault` set to true can exist in a cluster. If there is no PriorityClass defined with globalDefault set to true, all the pods with no priorityClassName defined will be treated with 0 priority (i.e. the least priority).
  - `description`: A string with a meaningful value so that people know when to use this PriorityClass.
-->
- PriorityClass 的名称必须是有效的 DNS 子域名。
- 当你创建自己的 PriorityClass 时，名称不应以 `system-` 开头，因为这类名称是被 Kubernetes 本身保留的（例如，它们被用于两个内置的 PriorityClass）。
- 其绝对值应在 -1000000000 到 1000000000（10 亿）之间。
- 较大的数值由 PriorityClass 保留，例如 `system-cluster-critical`（此 Pod 对集群至关重要）以及 `system-node-critical`（节点严重依赖此 Pod）。
  `system-node-critical` 的优先级高于 `system-cluster-critical`，因为集群级别关键 Pod 只有在其运行的节点满足其所有节点级别关键要求时才能正常工作。
- 额外两个可选字段：
  - `globalDefault`：当为 true 时，此 PriorityClass 用于未指定 `priorityClassName` 的 Pod。
    集群中只能存在一个 `globalDefault` 设置为 true 的 PriorityClass。
    如果没有 PriorityClass 的 globalDefault 设置为 true，则所有未定义 priorityClassName 的 Pod 都将被视为 0 优先级（即最低优先级）。
  - `description`：具备有意义值的字符串，以便人们知道何时使用此 PriorityClass。

> <!--
> {{< note >}}
> Adding a PriorityClass with `globalDefault` set to `true` does not mean it will apply the same to the existing pods that are already running. This will be applicable only to the pods that came into existence after the PriorityClass was created.
> {{< /note >}}
> -->
> {{< note >}}
> 添加一个将 `globalDefault` 设置为 `true` 的 PriorityClass 并不意味着它将同样应用于已在运行的现有 Pod。
> 这仅适用于创建 PriorityClass 之后出现的 Pod。
> {{< /note >}}

Reviewer 2: ```suggestion
{{< note >}}
<!--
Adding a PriorityClass with `globalDefault` set to `true` does not mean it will apply the same to the existing pods that are already running. This will be applicable only to the pods that came into existence after the PriorityClass was created.
-->
添加一个将 `globalDefault` 设置为 `true` 的 PriorityClass 并不意味着它将同样应用于已在运行的现有 Pod。
这仅适用于创建 PriorityClass 之后出现的 Pod。
{{< /note >}}
```

Author: Fixed in the latest update.

<!--
### PriorityClass in action
-->
### PriorityClass 的实际应用 {#priorityclass-in-action}

<!--
Here's an example. Next, create some environment-specific PriorityClasses:
-->
这里有一个例子。接下来，创建一些针对环境的 PriorityClass：

```yaml
apiVersion: scheduling.k8s.io/v1
kind: PriorityClass
metadata:
  name: dev-pc
value: 1000000
globalDefault: false
description: >-

>   (Optional) This priority class should only be used for all development pods.

Reviewer: This description could be in Chinese.

Author: Translated in the latest update.

```

```yaml
apiVersion: scheduling.k8s.io/v1
kind: PriorityClass
metadata:
  name: preprod-pc
value: 2000000
globalDefault: false
description: >-

>   (Optional) This priority class should only be used for all preprod pods.

Reviewer: This description could be in Chinese.

Author: Translated in the latest update.

```

```yaml
apiVersion: scheduling.k8s.io/v1
kind: PriorityClass
metadata:
  name: prod-pc
value: 4000000
globalDefault: false
description: >-

>   (Optional) This priority class should only be used for all prod pods.

Reviewer: This description could be in Chinese. Or it could be a mix of two languages.

Author: > This description could be in Chinese. Or it could be a mix of two languages.

Got it, I will fix all of these ASAP, not translate because some communities don't like to translate the script or code content.

-->
Kubernetes 已被广泛使用，许多组织将其用作事实上的编排引擎，用于运行需要频繁被创建和删除的工作负载。

<!--
Therefore, proper scheduling of the pods is key to ensuring that application pods are up and running within the Kubernetes cluster without any issues. This article delves into the use cases around resource management by leveraging the [PriorityClass](/docs/concepts/scheduling-eviction/pod-priority-preemption/#priorityclass) object to protect mission-critical or high-priority pods from getting evicted and making sure that the application pods are up, running, and serving traffic.
-->
因此，是否能正确对 Pod 进行调度是确保应用 Pod 在 Kubernetes 集群中正常启动并运行的关键。
本文深入探讨了围绕资源管理的使用场景，利用 [PriorityClass](/zh-cn/docs/concepts/scheduling-eviction/pod-priority-preemption/#priorityclass)
对象来保护关键或高优先级 Pod 免遭驱逐并确保应用 Pod 正常启动、运行以及提供流量。

<!--
## Resource management in Kubernetes
-->
## Kubernetes 中的资源管理 {#resource-management-in-kubernetes}

<!--
The control plane consists of multiple components, out of which the scheduler (usually the built-in [kube-scheduler](/docs/concepts/scheduling-eviction/kube-scheduler/)) is one of the components which is responsible for assigning a node to a pod.
-->

> 控制平面由多个组件组成，其中调度程序（通常是内置的 [kube-scheduler](/zh-cn/docs/concepts/scheduling-eviction/kube-scheduler/)）是一个负责为 Pod 分配节点的组件。

Reviewer 2: ```suggestion
控制平面由多个组件组成，其中调度程序（通常是内置的 [kube-scheduler](/zh-cn/docs/concepts/scheduling-eviction/kube-scheduler/)
是一个负责为 Pod 分配节点的组件。
```
虽然 en 没折行，但 zh 还是要 follow style guide 折行的额

<!--
Whenever a pod is created, it enters a "pending" state, after which the scheduler determines which node is best suited for the placement of the new pod.
-->
当 Pod 被创建时，它就会进入“Pending”状态，之后调度程序会确定哪个节点最适合放置这个新 Pod。

<!--
In the background, the scheduler runs as an infinite loop looking for pods without a `nodeName` set that are [ready for scheduling](/docs/concepts/scheduling-eviction/pod-scheduling-readiness/). For each Pod that needs scheduling, the scheduler tries to decide which node should run that Pod.
-->

> 在后台，调度程序以无限循环的方式运行，并寻找没有设置 `nodeName` 且[准备好进行调度](/zh-cn/docs/concepts/scheduling-eviction/pod-scheduling-readiness/)的 Pod。

Reviewer 2: ```suggestion
在后台，调度程序以无限循环的方式运行，并寻找没有设置 `nodeName`
且[准备好进行调度](/zh-cn/docs/concepts/scheduling-eviction/pod-scheduling-readiness/)的 Pod。
```

- Its absolute value should be between -1000000000 to 1000000000 (1 billion).
- Larger numbers are reserved by PriorityClasses such as `system-cluster-critical` (this Pod is critically important to the cluster) and `system-node-critical` (the node critically relies on this Pod). `system-node-critical` is a higher priority than `system-cluster-critical`, because a cluster-critical Pod can only work well if the node where it is running has all its node-level critical requirements met.
- There are two optional fields:
  - `globalDefault`: When true, this PriorityClass is used for pods where a `priorityClassName` is not specified. Only one PriorityClass with `globalDefault` set to true can exist in a cluster. If there is no PriorityClass defined with globalDefault set to true, all the pods with no priorityClassName defined will be treated with 0 priority (i.e. the least priority).
  - `description`: A string with a meaningful value so that people know when to use this PriorityClass.
-->
- PriorityClass 的名称必须是有效的 DNS 子域名。

> - 当你创建自己的 PriorityClass 时，名称不应以 `system-` 开头，因为这类名称是被 Kubernetes 本身保留的（例如，它们被用于两个内置的 PriorityClass）。

Reviewer 2: ```suggestion
- 当你创建自己的 PriorityClass 时，名称不应以 `system-` 开头，因为这类名称是被 Kubernetes
  本身保留的（例如，它们被用于两个内置的 PriorityClass）。
```

-->
[Pod 抢占](/zh-cn/docs/concepts/scheduling-eviction/pod-priority-preemption/#preemption)是 Kubernetes 的一项功能，
允许集群基于优先级抢占 Pod（删除现有 Pod 以支持新 Pod）。
[Pod 优先级](/zh-cn/docs/concepts/scheduling-eviction/pod-priority-preemption/#pod-priority)表示调度时 Pod 相对于其他 Pod 的重要性。
如果没有足够的资源来运行当前所有 Pod，调度程序会尝试驱逐优先级较低的 Pod，而不是优先级高的 Pod。

<!--
Also, when a healthy cluster experiences a node failure, typically, lower-priority pods get preempted to create room for higher-priority pods on the available node. This happens even if the cluster can bring up a new node automatically since pod creation is usually much faster than bringing up a new node.
-->
此外，当健康集群遇到节点故障时，通常情况下，较低优先级的 Pod 会被抢占，以便在可用节点上为较高优先级的 Pod 创建空间。
即使集群可以自动创建新节点，也会发生这种情况，因为 Pod 创建通常比创建新节点快得多。

<!--
### PriorityClass requirements
-->
### PriorityClass 的前提条件 {#priorityclass-requirements}

<!--
Before you set up PriorityClasses, there are a few things to consider.
-->

> 在设置 PriorityClass 之前，需要考虑一些事项。

Reviewer 3: ```suggestion
在配置 PriorityClass 之前，需要考虑一些事项。
```

NAME            READY   STATUS    RESTARTS   AGE   LABELS
dev-nginx       1/1     Running   0          55s   env=dev
preprod-nginx   1/1     Running   0          55s   env=preprod
prod-nginx      0/1     Pending   0          55s   env=prod
```

<!--
Bad news. The pod for the Production environment is still Pending and isn't serving any traffic.
-->
坏消息是生产环境的 Pod 仍处于 `Pending` 状态，并且不能提供任何流量。

<!--
Let's see why this is happening:
-->
让我们看看为什么会发生这种情况：

```console
$ kubectl get events
...
...
5s          Warning   FailedScheduling   pod/prod-nginx      0/2 nodes are available: 1 Insufficient cpu, 2 Insufficient memory.
```

<!--
In this example, there is only one worker node, and that node has a resource crunch.
-->
在此示例中，只有一个工作节点，并且该节点存在资源紧缩。

<!--
Now, let's look at how PriorityClass can help in this situation since prod should be given higher priority than the other environments.
-->
现在，让我们看看在这种情况下 PriorityClass 如何提供帮助，因为生产环境应该比其他环境具有更高的优先级。

<!--
## PriorityClass API
-->
## PriorityClass 的 API {#priorityclass-api}

<!--
Before creating PriorityClasses based on these requirements, let's see what a basic manifest for a PriorityClass looks like and outline some prerequisites:
-->

> 在根据这些需求创建 PriorityClass 之前，让我们看看 PriorityClass 的基本清单是什么样的，并提出一些先决条件：

Reviewer 3: ```suggestion
在根据这些需求创建 PriorityClass 之前，让我们看看 PriorityClass 的基本清单是什么样的，
并给出一些先决条件：
```

- There are two optional fields:
  - `globalDefault`: When true, this PriorityClass is used for pods where a `priorityClassName` is not specified. Only one PriorityClass with `globalDefault` set to true can exist in a cluster. If there is no PriorityClass defined with globalDefault set to true, all the pods with no priorityClassName defined will be treated with 0 priority (i.e. the least priority).
  - `description`: A string with a meaningful value so that people know when to use this PriorityClass.
-->
- PriorityClass 的名称必须是有效的 DNS 子域名。
- 当你创建自己的 PriorityClass 时，名称不应以 `system-` 开头，因为这类名称是被 Kubernetes
  本身保留的（例如，它们被用于两个内置的 PriorityClass）。
- 其绝对值应在 -1000000000 到 1000000000（10 亿）之间。
- 较大的数值由 PriorityClass 保留，例如 `system-cluster-critical`（此 Pod 对集群至关重要）以及 `system-node-critical`（节点严重依赖此 Pod）。
  `system-node-critical` 的优先级高于 `system-cluster-critical`，因为集群级别关键 Pod 只有在其运行的节点满足其所有节点级别关键要求时才能正常工作。
- 额外两个可选字段：

>   - `globalDefault`：当为 true 时，此 PriorityClass 用于未指定 `priorityClassName` 的 Pod。

Reviewer 3: ```suggestion
  - `globalDefault`：当为 true 时，此 PriorityClass 用于未设置 `priorityClassName` 的 Pod。
```

-->
dev-nginx Pod 即将被终止。一旦成功终止并且有足够的资源用于 prod Pod，控制平面就可以对 prod Pod 进行调度：

```console
Warning   FailedScheduling   pod/prod-nginx    0/2 nodes are available: 1 Insufficient cpu, 2 Insufficient memory.
Normal    Preempted          pod/dev-nginx     by default/prod-nginx on node node01
Normal    Killing            pod/dev-nginx     Stopping container dev-nginx
Normal    Scheduled          pod/prod-nginx    Successfully assigned default/prod-nginx to node01
Normal    Pulling            pod/prod-nginx    Pulling image "nginx"
Normal    Pulled             pod/prod-nginx    Successfully pulled image "nginx"
Normal    Created            pod/prod-nginx    Created container prod-nginx
Normal    Started            pod/prod-nginx    Started container prod-nginx
```

<!--
## Enforcement
-->
## 执行 {#enforcement}

<!--
When you set up PriorityClasses, they exist just how you defined them. However, people (and tools) that make changes to your cluster are free to set any PriorityClass, or to not set any PriorityClass at all. However, you can use other Kubernetes features to make sure that the priorities you wanted are actually applied.
-->

> 当你设置 PriorityClass 时，它们会按照你定义的方式存在。但是，对集群进行变更的人员（和工具）可以自由设置任意 PriorityClass，

Reviewer 3: ```suggestion
配置 PriorityClass 时，它们会按照你所定义的方式存在。
但是，对集群进行变更的人员（和工具）可以自由设置任意 PriorityClass，
```

> 或者根本不设置任何 PriorityClass。然而，你可以使用其他 Kubernetes 功能来确保实际应用你想要的优先级。

Reviewer 3: ```suggestion
或者根本不设置任何 PriorityClass。然而，你可以使用其他 Kubernetes 功能来确保你想要的优先级被实际应用起来。
```

<!--
As an alpha feature, you can define a [ValidatingAdmissionPolicy](/blog/2022/12/20/validating-admission-policies-alpha/) and a ValidatingAdmissionPolicyBinding so that, for example, Pods that go into the `prod` namespace must use the `prod-pc` PriorityClass. With another ValidatingAdmissionPolicyBinding you ensure that the `preprod` namespace uses the `preprod-pc` PriorityClass, and so on. In *any* cluster, you can enforce similar controls using external projects such as [Kyverno](https://kyverno.io/) or [Gatekeeper](https://open-policy-agent.github.io/gatekeeper/), through validating admission webhooks.
-->
作为一项 Alpha 级别功能，你可以定义一个 [ValidatingAdmissionPolicy](/blog/2022/12/20/validating-admission-policies-alpha/)
和一个 ValidatingAdmissionPolicyBinding，例如，进入 `prod` 命名空间的 Pod 必须使用 `prod-pc` PriorityClass。
通过另一个 ValidatingAdmissionPolicyBinding，你可以确保 `preprod` 命名空间使用 `preprod-pc` PriorityClass，依此类推。

> 在*任何*集群中，你可以使用外部项目，例如 [Kyverno](https://kyverno.io/) 或 [Gatekeeper](https://open-policy-agent.github.io/gatekeeper/) 通过验证准入网络钩子实施类似的控制。

Reviewer 3: ```suggestion
在*任何*集群中，你可以使用外部项目，例如 [Kyverno](https://kyverno.io/) 或 [Gatekeeper](https://open-policy-agent.github.io/gatekeeper/) 通过验证准入 Webhook 实施类似的控制。
```

<!--
However you do it, Kubernetes gives you options to make sure that the PriorityClasses are used how you wanted them to be, or perhaps just to [warn](https://open-policy-agent.github.io/gatekeeper/website/docs/violations/#warn-enforcement-action) users when they pick an unsuitable option.
-->

> 无论你如何操作，Kubernetes 都会为你提供用于确保按照你希望的方式使用 PriorityClass 的选项，

Reviewer 3: ```suggestion
无论你如何操作，Kubernetes 都会为你提供选项，确保 PriorityClass 的用法如你所愿，
```

或者只是当用户选择不合适的选项时做出[警告](https://open-policy-agent.github.io/gatekeeper/website/docs/violations/#warn-enforcement-action)。

<!--
## Summary
-->
## 总结 {#summary}

<!--
The above example and its events show you what this feature of Kubernetes brings to the table, along with several scenarios where you can use this feature. To reiterate, this helps ensure that mission-critical pods are up and available to serve the traffic and, in the case of a resource crunch, determines cluster behavior.
-->
上面的示例及其事件向你展示了 Kubernetes 此功能带来的好处，以及可以使用此功能的几种场景。

> 重申一下，这有助于确保关键任务 Pod 启动并可用于提供流量，并在资源紧张的情况下确定集群行为。

Reviewer 3: ```suggestion
重申一下，这一机制有助于确保关键任务 Pod 启动并可用于提供流量，并在资源紧张的情况下确定集群行为。
```
